File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.cpp
class Solution {
public:
    int findTheLongestSubstring(string s) {
        int n = s.size();

        map<char, int> mpp;
        mpp['a'] = 0;
        mpp['e'] = 1;
        mpp['i'] = 2;
        mpp['o'] = 3;
        mpp['u'] = 4;

        set<char> vowels = {'a', 'e', 'i', 'o', 'u'};

        int xr = 0;
        int res = 0;
        map<int, int> prevLocation;
        prevLocation[0] = -1;
        for (int i = 0; i < n; i++) {
            char ch = s[i];

            if (vowels.find(ch) != vowels.end()) {
                xr = xr ^ (1 << mpp[ch]);
            }

            if (prevLocation.find(xr) != prevLocation.end()) {
                res = max(res, i - prevLocation[xr]);
            } else {
                prevLocation[xr] = i;
            }
            
        }

        return res;
    }
};

File: 1371-find-the-longest-substring-containing-vowels-in-even-counts/1371-find-the-longest-substring-containing-vowels-in-even-counts.java
class Solution {
    public int findTheLongestSubstring(String s) {
        Map<Integer, Integer> firstOccurrence = new HashMap<>();

        int[] currentState = {0, 0, 0, 0, 0};
        firstOccurrence.put(getIntegerState(currentState), -1);

        Map<Character, Integer> idxMap = new HashMap<>();
        idxMap.put('a', 0);
        idxMap.put('e', 1);
        idxMap.put('i', 2);
        idxMap.put('o', 3);
        idxMap.put('u', 4);

        int ans = 0;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u')  {
                int idx = idxMap.get(ch);
                currentState[idx] = currentState[idx] ^ 1;
            }

            int integerState = getIntegerState(currentState);

            if (firstOccurrence.containsKey(integerState)) {
                ans = Math.max(ans, i - firstOccurrence.get(integerState));
            } else {
                firstOccurrence.put(integerState, i);
            }
        }

        return ans;
    }

    private int getIntegerState(int[] currentState) {
        int integerState = 0;
        for (int i = 0; i < 5; i++) {
            integerState = integerState + (int) Math.pow(10, i) * currentState[i];
        }

        return integerState;
    }
}